### ISSUE TITLE
How should -zapwallettxes interact with persistent mempool?

### ISSUE THREAD

**Issue by jnewbery** [bitcoin/bitcoin#9710]
the `zapwallettxes.py` qa test is currently broken due to a syntax error:

```python
     assert_raises(JSONRPCException, self.nodes[0].gettransaction, [txid3])
     #there must be a expection because the unconfirmed wallettx0 must be gone by now
```

This doesn't actually test that the transaction is not in the wallet, because `gettransaction()` is being called incorrectly (with an array instead of a string).

Changing the test to `assert_raises(JSONRPCException, self.nodes[0].gettransaction, txid3)` causes the rpc to succeed and the test script to fail because `txid3` is still in the wallet.

This is because the mempool is now persistent and is saved to disk when bitcoind shuts down. zapwallettxes removes the unconfirmed transaction from the wallet but bitcoind will rescan the mempool.dat file on startup and add the unconfirmed transaction back into the wallet.

I've 'fixed' zapwallettxes.py in https://github.com/jnewbery/bitcoin/tree/zapwallettxes , but the question remains: how *should* -zapwallettxes interact with persistent mempool? The use case for zapwallettxes is explained here: https://github.com/bitcoin/bitcoin/pull/3659#issuecomment-34878999 . My view is that even with zapwallettxes, bitcoind should rescan the mempool, but there should be better documentation or a warning that unconfirmed transaction in your mempool will be re-added to the wallet even if zapwallettxes is used. If you really want to remove all unconfirmed transactions from your wallet, you can delete your mempool.dat file and then restart with zapwallettxes.

**Comment by luke-jr** [comment]
Maybe we should have an option `-mempoolerase` or something.

**Comment by jnewbery** [comment]
@luke-jr yes that might be helpful.

To start with I'll add a warning message when someone starts bitcoind with -zapwallettxes and there's a mempool.dat file in their data directory. Something like:

"bitcoind started with -zapwallettxes, but there is a mempool file in the data directory. Unconfirmed transactions in the mempool file may be scanned and re-added to the wallet. To remove all unconfirmed transactions, you may wish to delete the mempool.dat file from your data directory."

If users are finding it too onerous to remove the mempool.dat file manually we could add a `-mempoolerase` option.

Sound good?

**Comment by jonasschnelli** [comment]
What about just not loading (=deleting/resetting) the `mempool.dat`-dump when `-zapwallettxes=<n>` was passed? Mention the new behaviour in the -zapwallettxes help/documentation.

**Comment by jnewbery** [comment]
@jonasschnelli you have much more experience of the wallet than I do. Not reloading the mempool when -zapwallettxes is set sounds like reasonable behaviour.

Any other suggestions before I prepare a fix? @sipa you introduced the persistent mempool. @laanwj you provided the use case for zapwallettxes in #3659. Does not loading mempool when zapwallettxes is set sound like reasonable behaviour?

**Comment by laanwj** [comment]
Yes, that may make sense. I'd implement it like this:
- Add a new option to delete/ignore the mempool at startup.
- Change parameter interaction so that `-zapwallettxes` implies that option.

This leaves flexibility to do othewise, avoids a tight coupling between a wallet and mempool option, but makes the default case to get rid of transactions: which is generally what people want when zapping the transactions.

**Comment by jnewbery** [comment]
#9966 was a pre-req for fixing this. It's now been merged so I've opened a PR to fix this issue at #10330

**Comment by Xekyo** [comment]
I just wanted to mention that a user on Bitcoin.Stackexchange complained that -zapwallettxes didn't work as expected due to the mempool reload reintroducing the transaction to his wallet, and that a default behavior of `-zapwallettxes` at least blocking the own transactions from being loaded would be expected. :)
